### PR TITLE
Python: Add hybrid search on text memory skill and memory store

### PR DESCRIPTION
**Pull request by jungwookim**
### Motivation and Context

  1. Why is this change required? It would be opening the various search types of memory.
  2. What problem does it solve? Currently, sk supports the nearest vector search.
  3. What scenario does it contribute to? This PR supports hybrid search.
  4. If it fixes an open issue, please link to the issue here. #3290 

### Description

I added the "search_type" signature so that we're able to select the search type such as "near_vector" or "hybrid" which is weaviate's a feature
.
### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by alexchaomander** [comment]
Thanks for adding this @jungwookim! I definitely agree making hybrid search available and easy to use is important! This might be good to get into main in the interim while the Kernel Memory project (https://github.com/microsoft/kernel-memory) works on bringing up a Python client that can be easily used by SK. 

@dluc what do you think?

File: python/semantic_kernel/core_skills/text_memory_skill.py [hunk 1]
```diff
@@ -39,6 +41,11 @@ class TextMemorySkill(PydanticField):
         description="The maximum number of relevant memories to recall.",
         default_value=DEFAULT_LIMIT,
     )
+    @sk_function_context_parameter(
+        name="search_type",
+        description="The type of search to perform: 'nearVector' or 'hybrid'",
```
Comment (alexchaomander) [inline python/semantic_kernel/core_skills/text_memory_skill.py:46]: I think naming-wise we can just call this `vector` search.

File: python/semantic_kernel/memory/semantic_text_memory.py [hunk 2]
```diff
@@ -133,6 +133,8 @@ async def search_async(
         limit: int = 1,
         min_relevance_score: float = 0.0,
         with_embeddings: bool = False,
+        search_type: str = "near_vector",
```
Comment (alexchaomander) [inline python/semantic_kernel/memory/semantic_text_memory.py:136]: rename to just `vector`

File: python/semantic_kernel/memory/semantic_text_memory.py [hunk 3]
```diff
@@ -149,13 +151,23 @@ async def search_async(
...
             await self._embeddings_generator.generate_embeddings_async([query])
         )[0]
-        results = await self._storage.get_nearest_matches_async(
-            collection_name=collection,
-            embedding=query_embedding,
-            limit=limit,
-            min_relevance_score=min_relevance_score,
-            with_embeddings=with_embeddings,
-        )
+        if search_type == "near_vector":
```
Comment (alexchaomander) [inline python/semantic_kernel/memory/semantic_text_memory.py:154]: rename to `vector`

File: python/semantic_kernel/memory/semantic_text_memory_base.py [hunk 4]
```diff
@@ -82,6 +82,7 @@ async def search_async(
         query: str,
         limit: int = 1,
         min_relevance_score: float = 0.7,
+        search_type: str = "near_vector",
```
Comment (alexchaomander) [inline python/semantic_kernel/memory/semantic_text_memory_base.py:85]: rename to `vector`

File: python/semantic_kernel/memory/semantic_text_memory_base.py [hunk 5]
```diff
@@ -92,6 +93,7 @@ async def search_async(
             limit {int} -- The maximum number of results to return. (default: {1})
             min_relevance_score {float} -- The minimum relevance score to return. (default: {0.0})
             with_embeddings {bool} -- Whether to return the embeddings of the results. (default: {False})
+            search_type {str} -- The type of search to perform. (default: {"near_vector"})
```
Comment (alexchaomander) [inline python/semantic_kernel/memory/semantic_text_memory_base.py:96]: rename to `vector`

**Review by alexchaomander — COMMENTED** [review]
Suggested changes to replace the name of `near_vector` to just `vector`

**Comment by jungwookim** [comment]
Thanks for reviewing @alexchaomander ! I've resolved all your comments. `vector` is way better to read for code.

**Comment by moonbox3** [comment]
Hi @jungwookim, is this work still needed? I am closing this PR as it has been stale since Oct 2023. If the work is still required, please open a new PR against the latest code in main. Thank you.

**Pull request closed without merging**